dbms/src/Functions/FunctionsDateTime.h

```diff
@@ -2814,42 +2813,282 @@ class FunctionExtractMyDateTime : public IFunction
+    }
+
+    static Int64 extractDayHour(Int64 nano)
+    {
+        MyDuration duration(nano);
+        return signMultiplier(duration) * duration.hours();
+    }
+};
+
+struct ExtractMyDateTimeFromStringImpl
+{
+    static Int64 extractDayMicrosecond(String dtStr)
+    {
+        Field duration_field = parseMyDuration(dtStr);
+        Int64 result = 0;
+        if (duration_field.isNull())
+        {
+            // TODO: should return null
+            return 0;
+        }
+        MyDuration duration(duration_field.template safeGet<Int64>());
+        result = ExtractMyDurationImpl::extractDayMicrosecond(duration.nanoSecond());
+
+        Field datetime_field = parseMyDateTime(dtStr);
+        if (!datetime_field.isNull())
+        {
+            MyDateTime datetime(datetime_field.template safeGet<UInt64>());
+            if (datetime.hour == duration.hours() && datetime.minute == duration.minutes() && datetime.second == duration.seconds() && datetime.year > 0)
             {
-                StringRef string_ref(data.data(), offsets[0] - 1);
-                constantString<F>(string_ref, from->size(), vec_to);
+                return ExtractMyDateTimeImpl::extractDayMicrosecond(datetime.toPackedUInt());
             }
-            else
+        }
+        return result;
+    }
+
+    static Int64 extractDaySecond(String dtStr)
+    {
+        Field duration_field = parseMyDuration(dtStr);
+        Int64 result = 0;
+        if (duration_field.isNull())
+        {
+            // TODO: should return null
+            return 0;
+        }
+        MyDuration duration(duration_field.template safeGet<Int64>());
+        result = ExtractMyDurationImpl::extractDaySecond(duration.nanoSecond());
+
+        Field datetime_field = parseMyDateTime(dtStr);
+        if (!datetime_field.isNull())
+        {
+            MyDateTime datetime(datetime_field.template safeGet<UInt64>());
+            if (datetime.hour == duration.hours() && datetime.minute == duration.minutes() && datetime.second == duration.seconds() && datetime.year > 0)
             {
-                vectorString<F>(data, offsets, vec_to);
+                return ExtractMyDateTimeImpl::extractDaySecond(datetime.toPackedUInt());
             }
         }
-        else if (const auto * from = checkAndGetColumn<ColumnUInt64>(col_from.get()); from)
+        return result;
+    }
+
+    static Int64 extractDayMinute(String dtStr)
+    {
+        Field duration_field = parseMyDuration(dtStr);
+        Int64 result = 0;
+        if (duration_field.isNull())
         {
-            const auto & data = from->getData();
-            if (checkColumnConst<ColumnUInt64>(from))
+            // TODO: should return null
+            return 0;
+        }
+        MyDuration duration(duration_field.template safeGet<Int64>());
+        result = ExtractMyDurationImpl::extractDayMinute(duration.nanoSecond());
+
+        Field datetime_field = parseMyDateTime(dtStr);
+        if (!datetime_field.isNull())
```

Comment:
It seems that we prefer to return value which is parsed by `parseMyDateTime`, how about moving codes related with paring duration to the tail of codes that parse datetime?

Not only in this function but also all similar functions

Reply:
in ParseMyDuration, we did not consider multiple cases which may cause parse failure.

Reply:
reference:

https://github.com/pingcap/tidb/blob/6feaa527160499ca646f7def098e4ea788b53113/types/time.go#L1802-L1824

Reply:
> We prefer to return value that is equivalently parsed by `parseMyDuration` and `parseMyDateTime`

okk